src/topology/mod.rs

```diff
@@ -606,51 +606,57 @@ impl RunningTopology {
 
         if let Some(inputs) = inputs {
             for input in inputs {
-                if let Some(output) = self.outputs.get(input) {
+                if let Some(output) = self.outputs.get_mut(input) {
                     // This can only fail if we are disconnected, which is a valid situation.
-                    let _ = output.send(fanout::ControlMessage::Remove(name.to_string()));
+                    let _ = output
+                        .send(fanout::ControlMessage::Remove(name.to_string()))
+                        .await;
                 }
             }
         }
     }
 
-    fn setup_outputs(&mut self, name: &str, new_pieces: &mut builder::Pieces) {
-        let output = new_pieces.outputs.remove(name).unwrap();
+    async fn setup_outputs(&mut self, name: &str, new_pieces: &mut builder::Pieces) {
+        let mut output = new_pieces.outputs.remove(name).unwrap();
 
         for (sink_name, sink) in &self.config.sinks {
             if sink.inputs.iter().any(|i| i == name) {
                 // Sink may have been removed with the new config so it may not be present.
                 if let Some(input) = self.inputs.get(sink_name) {
-                    output
+                    let _ = output
                         .send(fanout::ControlMessage::Add(sink_name.clone(), input.get()))
-                        .expect("Components shouldn't be spawned before connecting them together.");
+                        .await;
                 }
             }
         }
         for (transform_name, transform) in &self.config.transforms {
             if transform.inputs.iter().any(|i| i == name) {
                 // Transform may have been removed with the new config so it may not be present.
                 if let Some(input) = self.inputs.get(transform_name) {
-                    output
+                    let _ = output
                         .send(fanout::ControlMessage::Add(
                             transform_name.clone(),
                             input.get(),
                         ))
-                        .expect("Components shouldn't be spawned before connecting them together.");
+                        .await;
                 }
             }
         }
 
         self.outputs.insert(name.to_string(), output);
     }
 
-    fn setup_inputs(&mut self, name: &str, new_pieces: &mut builder::Pieces) {
+    async fn setup_inputs(&mut self, name: &str, new_pieces: &mut builder::Pieces) {
         let (tx, inputs) = new_pieces.inputs.remove(name).unwrap();
 
         for input in inputs {
             // This can only fail if we are disconnected, which is a valid situation.
-            let _ =
-                self.outputs[&input].send(fanout::ControlMessage::Add(name.to_string(), tx.get()));
+            let _ = self
+                .outputs
+                .get_mut(&input)
+                .unwrap()
+                .send(fanout::ControlMessage::Add(name.to_string(), tx.get()))
+                .await;
```

Comment:
I suppose the intent is to expose that this can panic, while that's true I think that for this case it's better to be concise and use `[index]` since we can't do anything reasonable with `None` besides panic.

Reply:
Unfortunately, the reason for this is simpler: `send` now expects `&mut self`, but `HashMap` doesn't implement `IndexMut`.

See https://github.com/rust-lang/rust/issues/23448 and https://github.com/rust-lang/rfcs/issues/997 for more info on that.